Types don't recursively (?) get propagated to JSON

Diffs can be applied to https://github.com/bittide/bittide-hardware/pull/801.

```diff
diff --git a/bittide-instances/src/Bittide/Instances/Tests/RegisterWbC.hs b/bittide-instances/src/Bittide/Instances/Tests/RegisterWbC.hs
index c981b591..96aab986 100644
--- a/bittide-instances/src/Bittide/Instances/Tests/RegisterWbC.hs
+++ b/bittide-instances/src/Bittide/Instances/Tests/RegisterWbC.hs
@@ -110,6 +110,12 @@ data SoP
   | SoP2
   deriving (Generic, NFDataX, ShowX, Show, ToFieldType, BitPackC, BitPack)
 
+data Inner = Inner
+  { innerA :: BitVector 8
+  , innerB :: BitVector 16
+  }
+  deriving (Generic, NFDataX, ShowX, Show, ToFieldType, BitPackC, BitPack)
+
 -- | Example that exports a bunch of different types.
 manyTypesWb ::
   forall wordSize aw dom.
@@ -159,6 +165,7 @@ manyTypesWb = circuit $ \(mm, wb) -> do
     , wbP1
     , wbP2
     , wbP3
+    , wbOI0
     ] <-
     deviceWbC "ManyTypes" -< (mm, wb)
 
@@ -207,6 +214,8 @@ manyTypesWb = circuit $ \(mm, wb) -> do
   registerWbC_ hasClock hasReset (registerConfig "p2") initP2 -< (wbP2, Fwd noWrite)
   registerWbC_ hasClock hasReset (registerConfig "p3") initP3 -< (wbP3, Fwd noWrite)
 
+  registerWbC_ hasClock hasReset (registerConfig "oi") initOI0 -< (wbOI0, Fwd noWrite)
+
   idC
  where
   initWbS0 :: Signed 8
@@ -308,6 +317,11 @@ manyTypesWb = circuit $ \(mm, wb) -> do
   initP3 :: P3
   initP3 = P3 (P2 0xBADC 0x0F) 0xEE
 
+  -- We purposely do not have a register for the `Inner` type, so we test that
+  -- types get put into the memory map recursively.
+  initOI0 :: Maybe Inner
+  initOI0 = Just (Inner 0x16 0x24)
+
 sim :: IO ()
 sim = putStrLn simResult
```

gets me:

```
$ cabal build bittide-instances/ && ./cargo.sh build
Up to date
firmware-support > cargo build
   Compiling bittide-hal v0.1.0 (/home/martijn/code/bittide-hardware/firmware-support/bittide-hal)
error[E0412]: cannot find type `Inner` in this scope
   --> bittide-hal/src/shared/devices/many_types.rs:184:31
    |
184 |     pub fn oi(&self) -> Maybe<Inner> {
    |                               ^^^^^ not found in this scope
    |
help: you might be missing a type parameter
    |
9   | impl<Inner> ManyTypes {
    |     +++++++
```

The type "Inner" isn't present in the JSON. Strangely, this doesn't happen when I define my own `Maybe` type:

```haskell
data MyMaybe a = MyNothing | MyJust a
  deriving (Generic, NFDataX, ShowX, Show, ToFieldType, BitPackC, BitPack)

...

  initOI0 :: MyMaybe Inner
  initOI0 = MyJust (Inner 0x16 0x24)
```

--- a/bittide-instances/src/Bittide/Instances/Tests/RegisterWbC.hs
+++ b/bittide-instances/src/Bittide/Instances/Tests/RegisterWbC.hs
@@ -82,6 +82,12 @@ data SoP
   | SoP2
   deriving (Generic, NFDataX, ShowX, Show, ToFieldType, BitPackC, BitPack)
 
+data Inner = Inner
+  { innerA :: BitVector 8
+  , innerB :: BitVector 16
+  }
+  deriving (Generic, NFDataX, ShowX, Show, ToFieldType, BitPackC, BitPack)
+
 -- | Example that exports a bunch of different types.
 manyTypesWb ::
   forall wordSize aw dom.
@@ -121,6 +127,7 @@ manyTypesWb = circuit $ \(mm, wb) -> do
     , wbSop0
     , wbSop1
     , wbE0
+    , wbOI0
     ] <-
     deviceWbC "ManyTypes" -< (mm, wb)
 
@@ -157,6 +164,8 @@ manyTypesWb = circuit $ \(mm, wb) -> do
   registerWbC_ hasClock hasReset (registerConfig "sop1") initSop1 -< (wbSop1, Fwd noWrite)
 
   registerWbC_ hasClock hasReset (registerConfig "e0") initWbE0 -< (wbE0, Fwd noWrite)
+
+  registerWbC_ hasClock hasReset (registerConfig "oi") initOI0 -< (wbOI0, Fwd noWrite)
 
   idC
  where
@@ -234,6 +243,11 @@ manyTypesWb = circuit $ \(mm, wb) -> do
 
   initWbE0 :: Either (BitVector 8) (BitVector 16)
   initWbE0 = Left 8
+
+  -- We purposely do not have a register for the `Inner` type, see issue
+  -- https://github.com/bittide/bittide-hardware/issues/815.
+  initOI0 :: Maybe Inner
+  initOI0 = Just (Inner 0x16 0x24)
 
 sim :: IO ()
 sim = putStrLn simResult

--- a/clash-protocols-memmap/src/Protocols/MemoryMap/TypeCollect.hs
+++ b/clash-protocols-memmap/src/Protocols/MemoryMap/TypeCollect.hs
@@ -62,7 +62,8 @@ collectDefs fieldType = case fieldType of
       def = TypeDescription tyName (L.length args') inner
       variantInner =
         L.concat $ L.concatMap (\(_name, fields) -> L.map collectDefs $ snd <$> fields) variants
+      argDefs = L.concatMap collectDefs args'
      in
-      (tyName, def) : variantInner
+      (tyName, def) : variantInner <> argDefs
   TypeReference ty args' -> collectDefs ty <> L.concatMap collectDefs args'
   _ -> []

